tools/checked-c-convert/ProgramInfo.cpp

```diff
@@ -31,6 +31,163 @@ bool ProgramInfo::checkStructuralEquality(uint32_t V, uint32_t U) {
   return false;
 }
 
+bool ProgramInfo::link() {
+
+  // For every global symbol in all the global symbols that we have found
+  // go through and apply rules for whether they are functions or variables.
+
+  for (const auto &S : GlobalSymbols) {
+    // First, extract out the function symbols from S. 
+    std::set<GlobalFunctionSymbol*> funcs;
+    std::set<GlobalVariableSymbol*> vars;
+    for (const auto &U : S.second)
+      if (GlobalFunctionSymbol *K = dyn_cast<GlobalFunctionSymbol>(U))
+        funcs.insert(K);
+      else if (GlobalVariableSymbol *V = dyn_cast<GlobalVariableSymbol>(U))
+        vars.insert(V);
+
+    // Then, pairwise-iterate over each of the function symbols found for this 
+    // symbol. What we want to do is for a sequence of constraint variables on
+    // F1,F2, set the constraint variables for F1(v0,vi,vN) and F2(v0,vj,vN) to
+    // be equal to each other, i.e. to enter a series of constraints of the form
+    // vi == vj.
+    //
+    // For example, consider a function that has been forward declared named 
+    // d1 with the signature void d1(int **c); There will be two sets of constraint
+    // variables in the system, one at the site of forward declaration and one at
+    // the site of function definition, like this:
+    //
+    // void d1(int *q_0 * q_1 c);
+    //
+    // void d1(int *q_2 * q_3 c) { *c = 0; }
+    //
+    // What we want to do is set q_0 == q_2 and q_1 == q_3. 
+    // To do that, we need to get the constraints for each parmvar decl position
+    // individually and set them equal, pairwise. 
+    for (std::set<GlobalFunctionSymbol*>::iterator I = funcs.begin();
+      I != funcs.end(); ++I) {
+      std::set<GlobalFunctionSymbol*>::iterator J = I;
+      J++;
+      if (J != funcs.end()) {
+        std::set<uint32_t> &rVars1 = (*I)->getReturns();
+        std::set<uint32_t> &rVars2 = (*J)->getReturns();
+        assert(rVars1.size() == rVars2.size());
```

Comment:
In real-world code bases, there might be mismatches in types.   The tool should probably report a problem and become conservative when it sees them.